File: Visualizer/Source/Visualizer/Persistence/Entity.py
import sqlite3
from typing import List, Dict


class Entity(object):
    # database = None
    # database_file = ""

    def _execute_sql(self, statement: str):
        if self.database is None:
            self.database = sqlite3.connect(self.database_file)

        cursor = self.database.cursor()
        cursor.execute(statement)

        return cursor.fetchall()

    def _on_save(self):
        pass

    def __init__(self, database_file=""):
        # Entity.database_file = database_file
        self.database = None
        self.database_file = database_file

    def _add_column(self, table: str, column: str, attributes: str):
        return self._execute_sql(f"ALTER TABLE {table} ADD {column} {attributes};")

    def _create_table(self, table: str, columns_with_attributes: Dict[str, str]):
        return self._execute_sql(
            f"CREATE TABLE {table}({','.join([column_with_attributes + ' ' + columns_with_attributes[column_with_attributes] for column_with_attributes in columns_with_attributes])});")

    def _insert(self, table: str, row_column_values):
        if isinstance(row_column_values, Dict):
            row_column_values: List[Dict[str, str]] = [row_column_values]

        # Collect columns
        columns: List[str] = list()
        for column_values in row_column_values:
            for column_value in column_values:
                columns.append(column_value)
        columns = sorted(set(columns))

        # Collect values
        row_values: List[List[str]] = list()
        for column_values in row_column_values:
            insert_values = list()

            for column in columns:
                insert_values.append(column_values[column])

            row_values.append(insert_values)

        return self._execute_sql(
            f"INSERT INTO {table}({','.join(columns)}) VALUES({'),('.join([','.join(row) for row in row_values])});")

    def _select(self, table: str, columns: List[str], conditions: str = None, order: str = None):
        return self._execute_sql(
            f"SELECT {','.join(columns)} FROM {table}" + (f" WHERE {conditions}" if conditions is not None else "") + (
                f" ORDER BY {order}" if order is not None else "") + ";")

    def save(self):
        self._on_save()

File: EnergyManager/Library/Source/EnergyManager/Persistence/Entity.cpp
#include "./Entity.hpp"

#include "EnergyManager/Utility/Logging.hpp"

namespace EnergyManager {
	namespace Persistence {
		sqlite3* Entity::database_ = nullptr;

		std::vector<std::map<std::string, std::string>> Entity::rows_ = {};

		int Entity::callback(void* context, int columnCount, char** columnValues, char** columnNames) {
			rows_.clear();

			std::map<std::string, std::string> row;

			for(size_t columnIndex = 0u; columnIndex < columnCount; ++columnIndex) {
				row[columnNames[columnIndex]] = columnValues[columnIndex];
			}

			rows_.push_back(row);

			return 0;
		}

		void Entity::onSave() {
		}

		void Entity::setID(const unsigned long& id) {
			id_ = id;
		}

		void Entity::initialize(const std::string& databaseFile) {
			Utility::Logging::logDebug("Initializing database %s...", databaseFile.c_str());

			// Open a new database connection
			if(database_ == nullptr && sqlite3_open(databaseFile.c_str(), &database_)) {
				ENERGY_MANAGER_UTILITY_EXCEPTIONS_EXCEPTION("Cannot open database: " + std::string(sqlite3_errmsg(database_)));
			}
		}

		std::vector<std::map<std::string, std::string>> Entity::executeSQL(const std::string& statement, const std::string& file, const int& line) {
			// Execute the statement
			char* errorMessage = nullptr;
			int errorCode = sqlite3_exec(database_, statement.c_str(), callback, nullptr, &errorMessage);
			if(errorCode) {
				throw EnergyManager::Utility::Exceptions::Exception(
					"Could not execute SQL statement " + statement + ": " + (errorMessage == nullptr ? std::to_string(errorCode) : std::string(errorMessage)),
					file,
					line);
			}

			return rows_;
		}

		void Entity::addColumn(const std::string& table, const std::string& column, const std::string& attributes) {
			ENERGY_MANAGER_PERSISTENCE_ENTITY_EXECUTE_SQL("ALTER TABLE " + table + " ADD " + column + " " + attributes + ";");
		}

		void Entity::createTable(const std::string& table, const std::map<std::string, std::string>& columnsWithAttributes) {
			std::vector<std::string> columnsWithAttributesSerialized;
			std::transform(columnsWithAttributes.begin(), columnsWithAttributes.end(), std::back_inserter(columnsWithAttributesSerialized), [](const auto& column) -> std::string {
				return column.first + " " + column.second;
			});
			ENERGY_MANAGER_PERSISTENCE_ENTITY_EXECUTE_SQL("CREATE TABLE " + table + "(" + Utility::Text::join(columnsWithAttributesSerialized, ",") + ");");
		}

		void Entity::insert(const std::string& table, const std::vector<std::map<std::string, std::string>>& rowColumnValues) {
			if(rowColumnValues.empty()) {
				return;
			}

			// Collect columns
			std::set<std::string> columns;
			for(const auto& columnValues : rowColumnValues) {
				for(const auto& columnValue : columnValues) {
					columns.insert(columnValue.first);
				}
			}

			// Collect values
			std::vector<std::vector<std::string>> rowValues;
			for(auto& columnValues : rowColumnValues) {
				std::vector<std::string> insertValues;

				for(const auto& column : columns) {
					auto insertValue = columnValues.at(column);
					insertValues.push_back(insertValue);
				}

				rowValues.push_back(insertValues);
			}

			std::vector<std::string> insertRows;
			std::transform(rowValues.begin(), rowValues.end(), std::back_inserter(insertRows), [](const auto& item) {
				return Utility::Text::join(item, ",");
			});

			// Store the results
			ENERGY_MANAGER_PERSISTENCE_ENTITY_EXECUTE_SQL("INSERT INTO " + table + "(" + Utility::Text::join(columns, ",") + ") VALUES(" + Utility::Text::join(insertRows, "),(") + ");");
		}

		unsigned long Entity::insert(const std::string& table, const std::map<std::string, std::string>& columnValues) {
			insert(table, std::vector<std::map<std::string, std::string>> { columnValues });

			return sqlite3_last_insert_rowid(database_);
		}

		void Entity::select(const std::string& table, const std::vector<std::string>& columns, const std::string& conditions) {
			ENERGY_MANAGER_PERSISTENCE_ENTITY_EXECUTE_SQL("SELECT " + Utility::Text::join(columns, ",") + " FROM " + table + " WHERE " + conditions + " ;");
		}

		void Entity::createIndex(const std::string& table, const std::string& index, const std::vector<std::string>& columns, const bool& unique) {
			ENERGY_MANAGER_PERSISTENCE_ENTITY_EXECUTE_SQL(std::string("CREATE") + (unique ? " UNIQUE" : "") + " INDEX " + index + " ON " + table + "(" + Utility::Text::join(columns, ",") + ");");
		}

		std::string Entity::filterSQL(std::string sql) {
			const std::string find = "\"";
			size_t startIndex = sql.find(find);
			if(startIndex != std::string::npos) {
				sql.replace(startIndex, find.length(), "\"\"");
			}

			return sql;
		}

		unsigned long Entity::getID() const {
			return id_;
		}

		void Entity::save() {
			onSave();
		}
	}
}